### PR TITLE
:goal_net: maybe_error

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <tyler@picknik.ai>

### CODE REVIEW

File: test/result_tests.cpp [hunk 1]
```diff
@@ -267,6 +267,19 @@ TEST(ResultTests, StringToViewNoThrow) {
   }
 }
 
+TEST(ResultTests, MaybeError) {
+  // GIVEN two result<T> values
```
Comment (griswaldbrooks) [inline test/result_tests.cpp]:
```suggestion
  // GIVEN two Result<T> values
```

File: test/result_tests.cpp [hunk 2]
```diff
@@ -267,6 +267,19 @@ TEST(ResultTests, StringToViewNoThrow) {
...
+  const Result<double> a = 6.5;
+  const Result<int> b = make_unexpected(Unknown());
```
Comment (griswaldbrooks) [inline test/result_tests.cpp:273]:
```suggestion
  const Result<int> b = make_unexpected(Unknown{});
```

File: include/monad/result.hpp [hunk 3]
```diff
@@ -244,3 +245,37 @@ struct fmt::formatter<Result<T>> {
...
+ * @brief Filter function for testing if a result has an error
+ *
+ * @param exp The expected type to test
+ * @tparam T The value type
+ * @tparam E The error type
+ * @return if the expected has an error
+ */
+template <typename T, typename E>
+constexpr bool has_error(const expected<T, E>& exp) {
+  return (!exp);
```
Comment (griswaldbrooks) [inline include/monad/result.hpp]: Why the parens? I assume there's a reason

File: examples/maybe_error.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+Result<double> f1(double x) { return divide_4_by(x); }
+Result<double> f2(double x) { return divide_4_by(x); }
+Result<double> f3(double x) { return divide_4_by(x); }
+Result<double> f4(double x) { return divide_4_by(x); }
+
+Result<double> do_work() {
+  auto const a = make_result(2.4) | f1 | f2;
+  auto const b = make_result(0.0) | f3 | f4;
+
+  if (auto const error = maybe_error(a, b); error) {
```
Comment (griswaldbrooks) [inline examples/maybe_error.cpp:49]: It really *feels* like this should return a `Result`, and not a `std::optional`, then couldn't we write
```
if (auto const error = maybe_error(a, b); error) {
  return error;
}
```
or am I missing something?

**Review by griswaldbrooks — CHANGES_REQUESTED** [review]
Remind me why we want this again?

File: include/monad/result.hpp (hunk 3 above)
Comment (tylerjw) [reply include/monad/result.hpp]: no, I just did it out of habit, I can drop them.

File: examples/maybe_error.cpp (hunk 4 above)
Comment (tylerjw) [reply examples/maybe_error.cpp:49]: If this returned a Result<T> what would the T be?  and would it even be useful.  The original idea was that T would be a tuple of all the values... but I'm not sure that's more useful than an optional<E> as what you want to test for is if any of them failed and return that error.

File: test/result_tests.cpp (hunk 2 above)
Comment (tylerjw) [reply test/result_tests.cpp:273]: Unknown is a constexpr lambda that returns an object so it has to be called this way... one thing I've been wondering about is should I put this all in some namespace?

File: examples/maybe_error.cpp (hunk 4 above)
Comment (griswaldbrooks) [reply examples/maybe_error.cpp:49]: So there's no `Result` type like there is a `Status` type that isn't `StatusOr`?

Comment (tylerjw) [reply examples/maybe_error.cpp:49]: no... I was sort of copying the rust names with

Result<T> = expected<T, Error>;

**Pull request merged**